Introduce guidelines to setup Supabase with Android Kotlin including sample with Jetpack Compose (#14571)
supabase/supabase: apps/docs/pages/guides/getting-started/quickstarts/kotlin.mdx

          nutrition text null,
          constraint products_pkey primary key (productid),
        ) tablespace pg_default;

      ````

    </StepHikeCompact.Code>

  </StepHikeCompact.Step>

  <StepHikeCompact.Step step={2}>

    <StepHikeCompact.Details title="Create an Android app with Android Studio">
      Setup environment for Android Development following official document.
      Open Android Studio > New > New Android Project

    </StepHikeCompact.Details>

  </StepHikeCompact.Step>

  <StepHikeCompact.Step step={3}>

    <StepHikeCompact.Details title="Install the Supabase client library">
    Setup Supabase libraries requires some others libraries from ktor, so make sure all dependencies are added
      
    </StepHikeCompact.Details>

    <StepHikeCompact.Code>

      ```kotlin
        implementation("io.ktor:ktor-client-cio:1.6.0")
        implementation("io.github.jan-tennert.supabase:postgrest-kt-android:0.9.0-alpha-4")
        implementation "io.ktor:ktor-client-android:2.2.1"
        implementation "io.ktor:ktor-utils-jvm:1.6.0"
        implementation "io.ktor:ktor-client-logging-jvm:1.6.0"
        implementation("io.ktor:ktor-client-core:2.2.4")

>       ```

Author: Fantanstic! I updated all the places are using version with this approach

    </StepHikeCompact.Code>

  </StepHikeCompact.Step>

  <StepHikeCompact.Step step={4}>

    <StepHikeCompact.Details title="Initialize the Supabase client">
    You can create a whenever you need to perform some API call.
    However, it is recommended that we should provide this kind of instance by using dependency injection library like Hilt. So see the code below to provide them properly. This is better for testability and maintainability
    </StepHikeCompact.Details>

    <StepHikeCompact.Code>

      ```kotlin
      // Handy create instance when need
      val client = createSupabaseClient(
      supabaseUrl = "https://xyzcompany.supabase.co", 
      supabaseKey = "public-anon-key") {
          install(GoTrue)
          install(Postgrest)
          //install other modules
      }

>       ```

Author: Thanks Jan, just reformated it

    </StepHikeCompact.Code>

  </StepHikeCompact.Step>

  <StepHikeCompact.Step step={4}>

    <StepHikeCompact.Details title="Create data transfer object">
    </StepHikeCompact.Details>

    <StepHikeCompact.Code>
      ```kotlin

>       @Serializable

Author: Exactly. Just a small thing but I forgot. I updated the step to setup serialization

      data class ProductDto(
          @SerialName("productid")
          val productId: String,
          @SerialName("name")
          val name: String,
          @SerialName("description")
          val description: String,
          @SerialName("price")
          val price: Double,
          @SerialName("image")
          val image: String,
          @SerialName("category")
          val category: String,
          @SerialName("nutrition")
          val nutrition: String,
          @SerialName("_id")
          val _id: Int,
      )
      ```

    </StepHikeCompact.Code>

  </StepHikeCompact.Step>

  <StepHikeCompact.Step step={5}>

    <StepHikeCompact.Details title="Query data from the app">
      
    </StepHikeCompact.Details>

    <StepHikeCompact.Code>
      ```kotlin

>       override suspend fun getProducts() {

Author: Updated!

          constraint products_pkey primary key (productid),
        ) tablespace pg_default;

      ````

    </StepHikeCompact.Code>

  </StepHikeCompact.Step>

  <StepHikeCompact.Step step={2}>

    <StepHikeCompact.Details title="Create an Android app with Android Studio">
      Setup environment for Android Development following official document.
      Open Android Studio > New > New Android Project

    </StepHikeCompact.Details>

  </StepHikeCompact.Step>

  <StepHikeCompact.Step step={3}>

    <StepHikeCompact.Details title="Install the Supabase client library">
    Setup Supabase libraries requires some others libraries from ktor, so make sure all dependencies are added.
    Replace place holder of the dependency versions `$supabase_version` and `$ktor_version` with latest versions as below
    <div style={{ display: 'flex', flexDirection: 'row', alignItems: 'center' }}> 
      <a href="https://github.com/supabase-community/supabase-kt/releases" style={{ marginRight: '8px' }}> 
          <img src="https://img.shields.io/github/release/supabase-community/supabase-kt?label=supabase-kt stable" alt="latest stable supabase-kt version"/> 
      </a> 

>       <a href="https://github.com/supabase-community/supabase-kt/releases"> 

Reviewer: ```suggestion
      <a href="https://github.com/ktorio/ktor/releases"> 
```

Author: Btw, how do you find these badges straightforward? I tried searching but it took me lots of time and I had to replace some key here and there to update

Reviewer: If you go to [shields.io](https://shields.io/) you can use the search bar on the top and enter terms like `github` or `maven` and you'll find the right one.

Author: Ah yeah thanks Jan. Another way is going to their official repository then copy 😅

          <img src="https://img.shields.io/github/v/release/ktorio/ktor.svg?label=ktor" alt="latest ktor version"/> 
      </a> 
    </div> 

    </StepHikeCompact.Details>

    <StepHikeCompact.Code>
      ```kotlin
        implementation "io.github.jan-tennert.supabase:postgrest-kt-android:$supabase_version"
        implementation "io.ktor:ktor-client-android:$ktor_version"
        implementation "io.ktor:ktor-client-logging-jvm:$ktor_version"
        implementation "io.ktor:ktor-client-core:$ktor_version"
        implementation "io.ktor:ktor-client-cio:$ktor_version"
        implementation "io.ktor:ktor-utils-jvm:$ktor_version"
      ```

    </StepHikeCompact.Code>

  </StepHikeCompact.Step>

  <StepHikeCompact.Step step={4}>

    <StepHikeCompact.Details title="Install the serializable plugin">
    Open `build.gradle` (app), add serialization plugin to use annotation for data parsing

    </StepHikeCompact.Details>

    <StepHikeCompact.Code>
      ```kotlin
        plugins {
            ...

>             id 'org.jetbrains.kotlin.plugin.serialization' version '1.7.0'

Reviewer: Maybe replace `1.7.0` with a placeholder like `KOTLIN_VERSION` as the plugin version should always be the same as the kotlin version

            ...
        }
      ```

    </StepHikeCompact.Code>

  </StepHikeCompact.Step>

  <StepHikeCompact.Step step={5}>

    <StepHikeCompact.Details title="Initialize the Supabase client">
    You can create a whenever you need to perform some API call.
    However, it is recommended that we should provide this kind of instance by using dependency injection library like Hilt. So see the code below to provide them properly. This is better for testability and maintainability
    </StepHikeCompact.Details>

    <StepHikeCompact.Code>

      ```kotlin
      // Handy create instance when need

>       val client = createSupabaseClient(
>           supabaseUrl = "https://xyzcompany.supabase.co", 
>           supabaseKey = "public-anon-key") {
>               install(GoTrue)
>               install(Postgrest)
>               install(Storage)
>           }

Reviewer: ```suggestion
      val client = createSupabaseClient(
          supabaseUrl = "https://xyzcompany.supabase.co", 
          supabaseKey = "public-anon-key"
      ) {
          install(GoTrue)
          install(Postgrest)
          install(Storage)
      }
```

    </StepHikeCompact.Details>

    <StepHikeCompact.Code>

     ```sql SQL_EDITOR
      -- Create products table
      create table
        public.products (
          _id bigint generated by default as identity not null,
          productid text not null,
          name text null,
          description text null,
          price real null,
          image text null,
          category text null,
          nutrition text null,
          constraint products_pkey primary key (productid),
        ) tablespace pg_default;

      ````

    </StepHikeCompact.Code>

  </StepHikeCompact.Step>

  <StepHikeCompact.Step step={2}>

    <StepHikeCompact.Details title="Create an Android app with Android Studio">
      Setup environment for Android Development following official document.
      Open Android Studio > New > New Android Project

    </StepHikeCompact.Details>

  </StepHikeCompact.Step>

  <StepHikeCompact.Step step={3}>

    <StepHikeCompact.Details title="Install the Supabase client library">
    Setup Supabase libraries requires some others libraries from ktor, so make sure all dependencies are added.

>     Replace place holder of the dependency versions `$supabase_version` and `$ktor_version` with latest versions as below

Reviewer: ```suggestion
    Replace the dependency version placeholders `$supabase_version` and `$ktor_version` with the latest versions as shown below.
```

  <StepHikeCompact.Step step={6}>

    <StepHikeCompact.Details title="Create data transfer object">
    </StepHikeCompact.Details>

    <StepHikeCompact.Code>
      ```kotlin
      @Serializable
      data class ProductDto(
          @SerialName("productid")
          val productId: String,
          @SerialName("name")
          val name: String,
          @SerialName("description")
          val description: String,
          @SerialName("price")
          val price: Double,
          @SerialName("image")
          val image: String,
          @SerialName("category")
          val category: String,
          @SerialName("nutrition")
          val nutrition: String,
          @SerialName("_id")
          val _id: Int,
      )
      ```

    </StepHikeCompact.Code>

  </StepHikeCompact.Step>

  <StepHikeCompact.Step step={7}>

    <StepHikeCompact.Details title="Query data from the app">
      
    </StepHikeCompact.Details>

    <StepHikeCompact.Code>
      ```kotlin
      suspend fun getProducts() {
        val result = client.postgrest["products"]
            .select().decodeList<ProductDto>()
        // Handle result data for next step
      }
      ```

>     </StepHikeCompact.Code>

Reviewer: Maybe it would be good to add another step showing the full activity or something like that? Without, this function seems a bit "alone", as there is nothing really going on (see flutter guide)

Author: You are right. I made it "less alone" 😂. Things should be provided from when we get the data from layer to layer until they are displayed to the view.

          @SerialName("name")
          val name: String,
          @SerialName("description")
          val description: String,
          @SerialName("price")
          val price: Double,
          @SerialName("image")
          val image: String,
          @SerialName("category")
          val category: String,
          @SerialName("nutrition")
          val nutrition: String,
          @SerialName("_id")
          val _id: Int,
      )
      ```

    </StepHikeCompact.Code>

  </StepHikeCompact.Step>

  <StepHikeCompact.Step step={7}>

    <StepHikeCompact.Details title="Create domain object">
      This kind of object will be consumed by the view
    </StepHikeCompact.Details>

    <StepHikeCompact.Code>
      ```kotlin
      data class Product(
          val productId: String,
          val name: String,
          val description: String,
          val price: Double,
          val image: String,
          val category: String,
          val nutrition: String,
          val _id: Int,
      )
      ```

    </StepHikeCompact.Code>

  </StepHikeCompact.Step>

  <StepHikeCompact.Step step={8}>

    <StepHikeCompact.Details title="Query data from the app">

>       

Reviewer: As you are using this inside a ProductRepository, maybe describe this is an actual function in that class/interface

    </StepHikeCompact.Details>

    <StepHikeCompact.Code>
      ```kotlin
      suspend fun getProducts(): ProductDto {
        val result = client.postgrest["products"]
            .select().decodeList<ProductDto>()
        // Handle result data for next step
        return result
      }
      ```
    </StepHikeCompact.Code>

  </StepHikeCompact.Step>

  <StepHikeCompact.Step step={9}>
    <StepHikeCompact.Details title="Get data from ViewModel inside a coroutine scope">

>      

Reviewer: As you are using Hilt, you may need a description saying that you are using Hilt in this case for dependency injection. See *link* to learn more about Hilt

    Replace the dependency version placeholders `$supabase_version` and `$ktor_version` with the latest versions as shown below.
    <div style={{ display: 'flex', flexDirection: 'row', alignItems: 'center' }}> 
      <a href="https://github.com/supabase-community/supabase-kt/releases" style={{ marginRight: '8px' }}> 
          <img src="https://img.shields.io/github/release/supabase-community/supabase-kt?label=supabase-kt stable" alt="latest stable supabase-kt version"/> 
      </a> 
      <a href="https://github.com/ktorio/ktor/releases"> 
          <img src="https://img.shields.io/github/v/release/ktorio/ktor.svg?label=ktor" alt="latest ktor version"/> 
      </a> 
    </div> 

    </StepHikeCompact.Details>

    <StepHikeCompact.Code>
      ```kotlin
        implementation "io.github.jan-tennert.supabase:postgrest-kt-android:$supabase_version"
        implementation "io.ktor:ktor-client-android:$ktor_version"
        implementation "io.ktor:ktor-client-logging-jvm:$ktor_version"
        implementation "io.ktor:ktor-client-core:$ktor_version"
        implementation "io.ktor:ktor-client-cio:$ktor_version"

>         implementation "io.ktor:ktor-utils-jvm:$ktor_version"

Reviewer: Is the "utils" and "logging" ktor artifact getting used? Also there are two Android engines used here, `CIO` and `Android`. Same with the Tutorial dependencies. `-android` and `-jvm` is also not needed.

            private val postgrest: Postgrest,
        ) : ProductRepository {
          override suspend fun getProducts(): List<ProductDto> {
            val result = client.postgrest["products"]
                .select().decodeList<ProductDto>()
            // Handle result data for next step
            return result
          }
      }
      ```
    </StepHikeCompact.Code>

  </StepHikeCompact.Step>

   <StepHikeCompact.Step step={9}>
    <StepHikeCompact.Details title="Create a module to provide repository">
    For dependency injection, we will use Hilt. Learn more about [Dependency injection with Hilt](https://developer.android.com/training/dependency-injection/hilt-android).
    </StepHikeCompact.Details>
    <StepHikeCompact.Code>
      ```kotlin
      InstallIn(SingletonComponent::class)
      @Module
      abstract class RepositoryModule {
          @Binds
          abstract fun bindProductRepository(impl: ProductRepositoryImpl): ProductRepository
      }
      ```
    </StepHikeCompact.Code>

  </StepHikeCompact.Step>

  <StepHikeCompact.Step step={10}>
    <StepHikeCompact.Details title="Get data from ViewModel inside a coroutine scope">

>     To use repository in ViewModel, use `@Inject` annotation.

Reviewer: ```suggestion
    To use the repository in a ViewModel, use the `@Inject` annotation.
```
